File: shell_sort.js
function shellSort(arr){
    if(arr.length === 0) return arr;
    
    const n = arr.length;
    let gap = Math.floor(n / 2);
    
    while (gap > 0){
        for (let i = gap; i < n; i++){
            let temp = arr[i];
            let j = i;
            
            while (j >= gap && arr[j-gap] > temp){
                arr[j] = arr[j-gap];
                j -= gap;
            }
            // replace j with temp
            arr[j] = temp;
        }
        
        gap = Math.floor(gap / 2);
    }
    
    return arr;
}

let arr = [56,23,6,2,0,9,1,0,32];
console.log("Sorted: ", shellSort(arr));

File: shell_sort.py
def shell_sort(arr):
  if not arr:
    return arr
    
  n = len(arr)
  gap = n // 2
  
  while gap > 0:
    for i in range(gap, n):
      temp = arr[i]
      j = i

      while j >= gap and arr[j-gap] > temp:
        arr[j] = arr[j-gap]
        j -= gap

      arr[j] = temp

    gap //= 2

  return arr


  # Example usage
arr = [0, 0, 1, 2, 6, 9, 32, 23, 56]
print("Sorted array:", shell_sort(arr))
    
